### ISSUE TITLE
MSOutput: decide which custodial site to use for data placement

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#9841]
**Impact of the new feature**
MSOutput

**Is your feature request related to a problem? Please describe.**
It looks like the we have overlooked the custodial output data placement. Custodial means that data is subscribed to Tape/MSS storage endpoints.

Right now, the output data placement in MicroServices does a data placement against any disk RSE (either DDM or Rucio will decide what the final location will be). So we need to change it to actually perform a custodial data placement (where MS needs to decide the final Tape location for DDM; with luck, Rucio can define it itself).

**Describe the solution you'd like**
We need to go through the checkor code:
https://github.com/CMSCompOps/WmAgentScripts/blob/master/Unified/checkor.py#L1009

and make/extract an algorithm for how the custodial site is defined within Unified. Once we know how it's done, we can either implement it in MSOutput or create a better way of identifying the final location.

Note that the Unified configuration is also used to define what goes or not to Tape (and maybe a few other parameters).

In short, this issue requires three actions:
1. study how custodial location is selected by Unified - AND make a clear specification in this issue
2. implement those requirements in the MSOutput 
3. evaluate whether we really need to have standard non-custodial output data placement done by MSOutput (given that workflows are already assigned with NonCustodial parameters, where WMAgent takes care of the output data placement).

**Describe alternatives you've considered**
Given that WMAgents *can* perform the output data placement - if workflows are assigned as such - it might be that the current MSOutput implement could actually be dropped, and no nonCustodial data placement needs to be performed.

**Additional context**
none

**Comment by amaltaro** [comment]
@vlimant Jean-Roch, if you can spare a couple of minutes on this issue. Do you see any need for nonCustodial data placement to be done by Unified (and soon by MSOutput)? Given that workflows are already assigned with NonCustodial information - thus output gets placed by WMAgent.

**Comment by amaltaro** [comment]
The feedback I got from Eric was that sites need to create/assign tape families, or make all the directories a dataset will live in. This means that Tape rules might take a while until they get approved (is it possible they would get disapproved?).

Of course, it does not explain why unified has only 25 campaigns configured with custodial location (out of 150 campaigns)...

@todor-ivanov I'm afraid you will need to carry this discussion, agree on a final decision and get whatever comes out of it implemented in MSOutput.

**Comment by amaltaro** [comment]
As discussed over the #wmagent-dev slack channel. Eric has prepared a prototype API that would select a weighted random TAPE RSE, code can be found here: https://github.com/ericvaandering/CMSRucio/blob/weighted_rucio_choice/src/pick_rse.py

We are now discussing other final details and what else needs to be considered for this choice, but once we agree on a final logic, the idea is to add it to the WMCore Rucio wrapper:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Rucio/Rucio.py

**Comment by nsmith-** [comment]
If you take the existing logic and make the weight variable a configuration parameter for the MSOutput module, then you could start implementation right away. We can work on the rucio side to construct the appropriate weight.

**Comment by amaltaro** [comment]
@ericvaandering @nsmith- if we want to make this pick_rse function generic enough. How about we pass the data volume (bytes) as input to this function, such that it can evaluate whether the RSE it's going to select can host that data or not (as part of the internal logic)?

**Comment by ericvaandering** [comment]
Yeah I was going to make that an option

**Comment by amaltaro** [comment]
Reopening it because we still have to figure out how to deal (and where to transfer) with the custodial data placement in the pre-Rucio era.
For the Rucio DM, Eric has already provided (which is now merged) a method to pick a random site based on an rse attr.
